src/symmetric_crypto.cpp

```diff
@@ -128,21 +157,37 @@ class AESCipher::Impl
                 throw MoCOCrWException("Authentication Tag must be set before calling finish() "
                                        "for cipher which implements authenticated encryption.");
             }
-            openssl::_EVP_CIPHER_CTX_ctrl(_ctx.get(), EVP_CTRL_GCM_SET_TAG,
+            _EVP_CIPHER_CTX_ctrl(_ctx.get(), EVP_CTRL_GCM_SET_TAG,
                                           AuthenticatedAESCipher::AuthTagLength, _authTag.data());
         }
 
-        openssl::_EVP_CipherFinal_ex(_ctx.get(), processedChunk.data(), &processingChunkSize);
+        try {
+            _EVP_CipherFinal_ex(_ctx.get(), processedChunk.data(), &processingChunkSize);
+        }
+        catch (const OpenSSLException &e) {
+            // OpenSSL does not set any specific error codes which we can use to distinguish
+            // authentication failure from other type of errors. Therefore, if there is an error
+            // on decrypting authenticated ciphertext we assume that there is a problem with
+            // the key, padding or IV.
+            if (isAuthenticatedCipherMode(_mode) && _operation == Operation::Decryption) {
+                throw MoCOCrWException(
```

Comment:
maybe we still want to propagate openssl's error message?

Reply:
As comment explains (probably badly), OpenSSL does not provide any reasonable error code or error description in this case. It is nothing useful but confusing "file:null line:null.." type of error.